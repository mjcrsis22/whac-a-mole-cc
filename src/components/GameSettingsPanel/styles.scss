@use '../../scss/variables.scss' as *;

.game-settings-panel {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding: 0.5rem 1rem;
  background-image: url(/backgrounds/zig-zag-forest.svg);
  background-size: contain;
  background-color: #00bb77;
  color: white;

  .level-button {
    padding: 0.2rem 0.4rem;
    margin-left: 1.5rem;
    color: #00bb77;

    &:hover {
      border-color: darkslategray;
    }

    &.selected {
      background-color: darkslategray;
    }
  }

  @media (max-width: $layout-breakpoint-small) {
    flex-direction: column;
    .level-button {
      margin: 0.2rem;
    }
  }
}
