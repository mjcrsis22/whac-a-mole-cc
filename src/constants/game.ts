export enum GAME_LEVEL {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HARD = 'HARD',
}

export interface GameSetting {
  id: GAME_LEVEL;
  spotRate: number;
  points: number;
}

export const GAME_SETTINGS_BY_LEVEL: {
  [key in keyof typeof GAME_LEVEL]: GameSetting;
} = {
  [GAME_LEVEL.LOW]: {
    id: GAME_LEVEL.LOW,
    spotRate: 1000,
    points: 10,
  },
  [GAME_LEVEL.MEDIUM]: {
    id: GAME_LEVEL.MEDIUM,
    spotRate: 750,
    points: 20,
  },
  [GAME_LEVEL.HARD]: {
    id: GAME_LEVEL.HARD,
    spotRate: 500,
    points: 30,
  },
};
