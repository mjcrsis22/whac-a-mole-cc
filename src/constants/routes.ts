export enum APP_ROUTES {
  HOME = '/home',
  GAME = '/game',
}
