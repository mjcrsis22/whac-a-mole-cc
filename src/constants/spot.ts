export enum SPOT_STATES {
  inactive,
  active,
}

export const SPOT_POSITIONS = [
  { left: '5%', top: '0%' },
  { left: '5%', top: '72%' },
  { left: '30%', top: '15%' },
  { left: '30%', top: '60%' },
  { left: '45%', top: '40%' },
  { left: '60%', top: '15%' },
  { left: '60%', top: '60%' },
  { left: '85%', top: '0%' },
  { left: '85%', top: '72%' },
];
