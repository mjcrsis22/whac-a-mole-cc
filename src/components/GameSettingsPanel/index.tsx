import { useContext } from 'react';
import { GameContext } from '../../contexts/GameContext';
import { ActionTypes } from '../../contexts/GameContext/reducer';
import { GAME_LEVEL, GAME_SETTINGS_BY_LEVEL } from '../../constants/game';
import './styles.scss';

function GameSettingsPanel() {
  const {
    state: { levelSetting },
    dispatch,
  } = useContext(GameContext);

  const getSelectedClassName = (level: GAME_LEVEL) => {
    return level === levelSetting.id ? 'selected' : '';
  };
  const handleLevelClick = (level: GAME_LEVEL) => {
    return () =>
      dispatch({
        type: ActionTypes.SET_LEVEL,
        payload: GAME_SETTINGS_BY_LEVEL[level],
      });
  };

  return (
    <div className="game-settings-panel">
      <span>Levels</span>
      <div className="level-buttons">
        {Object.values(GAME_LEVEL).map(level => (
          <button
            key={level}
            className={`button level-button ${getSelectedClassName(level)}`}
            onClick={handleLevelClick(level)}
            title={`Level ${level}`}
          >
            {level.toUpperCase()}
          </button>
        ))}
      </div>
    </div>
  );
}

export default GameSettingsPanel;
