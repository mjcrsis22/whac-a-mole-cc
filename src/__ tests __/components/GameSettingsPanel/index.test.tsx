import { useContext } from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import GameContextProvider from '../../../contexts/GameContext';
import {
  ActionTypes,
  INITIAL_STATE,
} from '../../../contexts/GameContext/reducer';
import { GAME_LEVEL, GAME_SETTINGS_BY_LEVEL } from '../../../constants/game';
import GameSettingsPanel from '../../../components/GameSettingsPanel';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: jest.fn(),
}));

const mockedContext = {
  state: INITIAL_STATE,
  dispatch: jest.fn(),
};

const mockedUseContext = useContext as jest.MockedFn<typeof useContext>;

describe('GameSettingsPanel component', () => {
  beforeEach(() => {
    mockedUseContext.mockReturnValue(mockedContext);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('Renders correctly', () => {
    const { container } = render(
      <GameContextProvider>
        <GameSettingsPanel />
      </GameContextProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test('Dispatches SET_LEVEL action on button click', async () => {
    const result = render(
      <GameContextProvider>
        <GameSettingsPanel />
      </GameContextProvider>,
    );

    const buttons = result.getAllByRole('button');
    expect(buttons.length).toEqual(3);
    await userEvent.click(buttons[1]);

    expect(mockedContext.dispatch).toHaveBeenCalledWith({
      type: ActionTypes.SET_LEVEL,
      payload: GAME_SETTINGS_BY_LEVEL[GAME_LEVEL.MEDIUM],
    });
  });
});
