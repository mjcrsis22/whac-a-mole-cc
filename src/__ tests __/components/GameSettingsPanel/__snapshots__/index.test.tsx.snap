// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameSettingsPanel component Renders correctly 1`] = `
<div>
  <div
    class="game-settings-panel"
  >
    <span>
      Levels
    </span>
    <div
      class="level-buttons"
    >
      <button
        class="button level-button selected"
        title="Level LOW"
      >
        LOW
      </button>
      <button
        class="button level-button "
        title="Level MEDIUM"
      >
        MEDIUM
      </button>
      <button
        class="button level-button "
        title="Level HARD"
      >
        HARD
      </button>
    </div>
  </div>
</div>
`;
